python/lsst/rubintv/production/rubinTv.py

```diff
@@ -480,69 +520,106 @@ def dataIdToMetadataDict(butler, dataId, keysToRemove):
 
         for key in keysToRemove:
             if key in d:
                 d.pop(key)
 
-        return {seqNum: d}
+        properNames = {MD_NAMES_MAP[attrName]: d[attrName] for attrName in d}
 
-    @staticmethod
-    def appendToJson(filename, md):
-        """Add a dictionary item to the JSON file containing the sidecar data.
+        return {seqNum: properNames}
+
+    def writeShardForDataId(self, dataId):
+        """Write a standard shard for this dataId from the exposure record.
 
-        Updates the file in place.
+        Calls dataIdToMetadataDict to get the normal set of metadata components
+        and then writes it to a shard file in the shards directory ready for
+        upload.
 
         Parameters
         ----------
-        filename : `str`
-            The filename
-        md : `dict`
-            The metadata, as a dict, to add to the JSON file.
+        dataId : `dict`
+            The dataId.
         """
-        data = {}
-        if os.path.isfile(filename) and os.path.getsize(filename) > 0:  # json.load() doesn't like empty files
-            with open(filename) as f:
-                data = json.load(f)
-        data.update(md)
+        md = self.dataIdToMetadataDict(self.butler, dataId, SIDECAR_KEYS_TO_REMOVE)
+        dayObs = butlerUtils.getDayObs(dataId)
+        writeMetataShard(self.shardsDir, dayObs, md)
+        return
 
-        with open(filename, 'w') as f:
-            json.dump(data, f)
+    def mergeShardsAndUpload(self):
+        """Merge all the shards in the shard directory into their respective
+        files and upload the updated files.
 
-    def getSidecarFilename(self, dataId):
-        """Get the name of the metadata sidecar file for the dataId.
+        For each file found in the shard directory, merge its contents into the
+        main json file for the corresponding dayObs, and for each file updated,
+        upload it.
+        """
+        filesTouched = set()
+        shardFiles = sorted(glob(os.path.join(self.shardsDir, "metadata-*")))
+        if shardFiles:
+            self.log.debug(f'Found {len(shardFiles)} shardFiles')
+            sleep(0.1)  # just in case a shard is in the process of being written
+
+        for shardFile in shardFiles:
+            # filenames look like
+            # metadata-dayObs_20221027_049a5f12-5b96-11ed-80f0-348002f0628.json
+            filename = os.path.basename(shardFile)
+            dayObs = int(filename.split("_", 2)[1])
+            mainFile = self.getSidecarFilename(dayObs)
+            filesTouched.add(mainFile)
+
+            data = {}
+            # json.load() doesn't like empty files so check size is non-zero
+            if os.path.isfile(mainFile) and os.path.getsize(mainFile) > 0:
+                with open(mainFile) as f:
+                    data = json.load(f)
+
+            with open(shardFile) as f:
+                shard = json.load(f)
+            if shard:
+                for row in shard:
```

Comment:
Almost. So, `data` is the contents of the main file, which the shard's contents are being merged into - that's just been loaded from disk from `mainFile`. Each shard can be one or more rows (where a row is a seqNum), and each row is a dict of metadata with one or more entries. So the minimal shard is a single k,v pair for a single seqNum, but it could also be 20 k,v pairs for 150 seqNums, for example, and there could also be multiple shards, all being merged into the main file.

You are quite right about the potential for this automatically taking the last entry and dropping all others, but this is for filling gaps in tables and adding new entries. If we're re-writing items here something has gone desperately wrong with the other services and this is the least of our worries.